src/cpp/session/modules/SessionPath.cpp

```diff
@@ -39,129 +43,165 @@ namespace session {
-}
-
-void addToPathIfNecessary(
-      const std::string& entry,
-      std::vector<std::string>* pPathComponents)
+   {
+      return error;
+   }
+   else if (result.exitStatus != EXIT_SUCCESS)
+   {
+      return systemError(
+               boost::system::errc::state_not_recoverable,
+               result.stdErr,
+               ERROR_LOCATION);
+   }
+   else
+   {
+      // only include last line of output, in case the shell printed
+      // something extra during login / processing of scripts
+      std::string output = core::string_utils::trimWhitespace(result.stdOut);
+      std::vector<std::string> lines = core::algorithm::split(output, "\n");
+      auto n = lines.size();
+      if (n == 0)
+      {
+         return systemError(
+                  boost::system::errc::state_not_recoverable,
+                  result.stdErr,
+                  ERROR_LOCATION);
+      }
+      
+      // extract last line of output
+      std::string path = lines[n - 1];
+      pPath->assign(path);
+      return Success();
+   }
+}   
+
+// this routine is a little awkward -- if RStudio was launched from a terminal,
+// it's possible that the PATH is already set as appropriate for any program
+// launched through a shell. as a heuristic, we try to see if 'usr/local/bin'
+// is already on the PATH; if it is, we assume the path is appropriately
+// initialized; otherwise, we try to initialize it in the same way a shell
+// might.
+std::string initializePath()
 {
-   // tolerate paths with trailing slashes
-   for (std::string item : {entry, entry + "/"})
+   // if the user's path already contains '/usr/local/bin', assume that
+   // they're running RStudio through a shell / terminal and so we don't
+   // need to re-read the shell PATH
+   std::string defaultPath = core::system::getenv("PATH");
+   boost::regex reUsrLocalbin("(^|:)/usr/local/bin(:|$)");
+   if (boost::regex_search(defaultPath, reUsrLocalbin))
+      return defaultPath;
+   
+   // first, try to initialize with user's default shell
+   // (RSTUDIO_SESSION_SHELL is primarily for internal use)
+   std::string shell = core::system::getenv("RSTUDIO_SESSION_SHELL");
+   if (shell.empty())
+      shell = core::system::getenv("SHELL");
+   
+   if (!shell.empty())
+   {
+      std::string path;
+      Error error = initializePathViaShell(shell, &path);
+      if (!error)
+         return path;
+   }
+   
+   // next, try to initialize with default shell
+   if (shell != "/bin/sh")
    {
-      auto it = std::find(
-               pPathComponents->begin(),
-               pPathComponents->end(),
-               item);
-
-      // if we find the path component, bail (no need to add to PATH)
-      if (it != pPathComponents->end())
-         return;
+      std::string path;
+      Error error = initializePathViaShell("/bin/sh", &path);
+      if (!error)
+         return path;
    }
+   
+   // all else fails, use the current application path
+   return core::system::getenv("PATH");
```

Comment:
if this fails, don't we need to pull in paths from /etc? Thought the fallback would essentially be like your conservative PR 2

Reply:
That's true -- I should add that back in.